lib/Transforms/CSE.cpp

```diff
@@ -0,0 +1,783 @@
+//===- CSE.cpp - Common Sub-expression Elimination ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass performs a simple common sub-expression elimination
+// algorithm on operations within a region.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Transforms/Passes.h"
+
+#include "mlir/IR/Dominance.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/RecyclingAllocator.h"
+#include <deque>
+
+using namespace mlir;
+using namespace circt;
+
+namespace {
+
+/// ----- Start of LLVM upstream changes.
+
+/// TODO: Upstream following changes to mlir/IR/OperationSupport.h
+
+//===----------------------------------------------------------------------===//
+// Operation Equivalency
+//===----------------------------------------------------------------------===//
+
+struct OperationEquivalence {
+  enum Flags {
```

Comment:
It seems like in preparation to upstream this, custom attribute handling should be an equivalence flag.

Reply:
We just discussed this one. If we are adding a more general callback mechanism, maybe we can remove just remove the flags and instead provide callbacks upstream for the behaviors previously dictated by the flags.